mesop/components/html/html.ts

```diff
@@ -16,12 +17,52 @@ export class HtmlComponent {
   @Input({required: true}) type!: Type;
   @Input() key!: Key;
   @Input() style!: Style;
+  @ViewChild('iframe', {read: ElementRef}) iframe!: ElementRef;
   private _config!: HtmlType;
+  isSandboxed!: boolean;
+  html!: string;
 
   ngOnChanges() {
     this._config = HtmlType.deserializeBinary(
       this.type.getValue() as unknown as Uint8Array,
     );
+    this.isSandboxed = this._config.getMode() === 'sandboxed';
+    if (!this.isSandboxed) {
+      return;
+    }
+    const previousHtml = this.html;
+    this.html = this._config.getHtml()!;
+    // Reload iframe if the HTML has changed.
+    if (
+      this.html !== previousHtml &&
+      this.iframe &&
+      this.iframe.nativeElement
+    ) {
+      this.loadIframe();
+    }
+  }
+
+  ngAfterViewInit() {
+    this.loadIframe();
+  }
+
+  loadIframe(): void {
+    if (!this.iframe) {
+      if (this.isSandboxed) {
+        console.warn(
+          'iframe element in Mesop html component unexpectedly not found',
+        );
+      }
+      return;
+    }
+    const iframe = this.iframe.nativeElement as HTMLIFrameElement;
+    setIframeSrc(iframe, '/sandbox_iframe.html', {allowSameOrigin: false});
```

Comment:
[question] What's the reason for setting `allowSameOrigin` false in this case but not for embed? My assumption is that the sandbox file is on the Mesop server so we don't want it accessing things like cookies especially if the content will be dynamically generated based on user input. And I guess with the embed case, we assume the embed will typically be from a different origin and that site may need access to the cookies of their origin?

[question] I remember you mentioning before about using src doc attribute. But here I see you're loading this sandbox html file and then updating the HTML through the message event. I'm assuming this is a safer approach?

Reply:
Good questions. It's pretty complicated and I left detailed comments both here and in embed.

For this case (sandbox_iframe.html) it's important that we don't allow it to be on the same origin because user inputs are being executed inside that document with no sanitization, e.g. imagine a callsite like:

`me.html(me.state(State).input, mode="sandboxed")` in this case, we don't want to execute end-user input text which could be something like (go send an XHR to some other server with sensitive data), esp. in the case where an end-user could be tricked to put something malicious.

Right, for the embed case, it's usually a different origin, and many sites won't behave properly if you don't let it run on its original origin.

---

I initially tried using srcdoc however the problem is that it inherits the CSP of the main document (index.html) which has a very stringent CSP and would basically make this feature unusable. By using this sandbox_iframe.html shim layer, I can set a much more lax CSP (with the assumption that this page is carefully sandboxed when iframed).